// cucumber/lib/models/definition.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Definition = function () {
  function Definition(_ref) {
    var code = _ref.code,
        line = _ref.line,
        options = _ref.options,
        uri = _ref.uri;
    (0, _classCallCheck3.default)(this, Definition);

    this.code = code;
    this.line = line;
    this.options = options;
    this.uri = uri;
  }

  (0, _createClass3.default)(Definition, [{
    key: "buildInvalidCodeLengthMessage",
    value: function buildInvalidCodeLengthMessage(syncOrPromiseLength, callbackLength) {
      return "function has " + this.code.length + " arguments" + (", should have " + syncOrPromiseLength + " (if synchronous or returning a promise)") + (" or " + callbackLength + " (if accepting a callback)");
    }
  }, {
    key: "getInvalidCodeLengthMessage",
    value: function getInvalidCodeLengthMessage(parameters) {
      return this.buildInvalidCodeLengthMessage(parameters.length, parameters.length + 1);
    }
  }]);
  return Definition;
}();

exports.default = Definition;
