// cucumber/lib/models/test_case_hook_definition.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _pickle_filter = require('../pickle_filter');

var _pickle_filter2 = _interopRequireDefault(_pickle_filter);

var _definition = require('./definition');

var _definition2 = _interopRequireDefault(_definition);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TestCaseHookDefinition = function (_Definition) {
  (0, _inherits3.default)(TestCaseHookDefinition, _Definition);

  function TestCaseHookDefinition() {
    var _ref;

    (0, _classCallCheck3.default)(this, TestCaseHookDefinition);

    for (var _len = arguments.length, data = Array(_len), _key = 0; _key < _len; _key++) {
      data[_key] = arguments[_key];
    }

    var _this = (0, _possibleConstructorReturn3.default)(this, (_ref = TestCaseHookDefinition.__proto__ || Object.getPrototypeOf(TestCaseHookDefinition)).call.apply(_ref, [this].concat(data)));

    _this.pickleFilter = new _pickle_filter2.default({
      tagExpression: _this.options.tags
    });
    return _this;
  }

  (0, _createClass3.default)(TestCaseHookDefinition, [{
    key: 'appliesToTestCase',
    value: function appliesToTestCase(_ref2) {
      var pickle = _ref2.pickle,
          uri = _ref2.uri;

      return this.pickleFilter.matches({ pickle: pickle, uri: uri });
    }
  }, {
    key: 'getInvalidCodeLengthMessage',
    value: function getInvalidCodeLengthMessage() {
      return this.buildInvalidCodeLengthMessage('0 or 1', '2');
    }
  }, {
    key: 'getInvocationParameters',
    value: function getInvocationParameters(_ref3) {
      var hookParameter = _ref3.hookParameter;

      return [hookParameter];
    }
  }, {
    key: 'getValidCodeLengths',
    value: function getValidCodeLengths() {
      return [0, 1, 2];
    }
  }]);
  return TestCaseHookDefinition;
}(_definition2.default);

exports.default = TestCaseHookDefinition;
